### ISSUE TITLE
[Docs] Questions about square averaging

### ISSUE THREAD

**Issue by KaiWU5** [OpenGVLab/InternVL#746]
### 📚 The doc issue

![image](https://github.com/user-attachments/assets/91baf3c9-f670-4fed-b093-7859b39753b9)
Thanks for your fantastic work for cutting-edge InternVL2.5, I am curious the loss mentioned in the paper about "square averaging". What does the $x$ mean, when the equation has $x^1$? I think the $x = x_i / \sum{x}$  where $i$ is the ith sample, and x is the number of tokens in each sample. So that the $x$ means the average across samples not tokens

**Comment by lll2343** [comment]
Hi,

$x$ represents the total number of tokens in the sample (response) to which token $i$ belongs. It is not the average across multiple samples but is specific to the **sample** that token $i$ is part of. 

For example, if the sample is "what can I say" then $x = 4$ (as the sample contains 4 tokens). Based on this:
- $x^1 = x = 4$
- $x^0 = 1$
- $x^{0.5} = \sqrt{x} = 2$

These powers of $x$ are used in different weighting strategies, such as token averaging ($x^0$), sample averaging ($x^1$), and square averaging ($x^{0.5}$).

**Comment by KaiWU5** [comment]
Thanks a lot for @lll2343 reply and in-depth explanation. I put my code according to the paper to put everyone else on the same page
```python
# count for average
# elements not equal to -100
not_minus_100 = shift_labels.ne(-100)
# count of elements not equal to -100 in each 1D tensor along the 0th dimension
x_counts = not_minus_100.sum(dim=1)
square_weights_by_sample = torch.square(x_counts / x_counts.sum())

# Flatten the tokens
shift_logits = shift_logits.view(-1, vocab_size)
shift_labels = shift_labels.view(-1)
# Enable square weighting
shift_labels = shift_labels.to(shift_logits.device)
loss = loss_fct(shift_logits, shift_labels)
loss = loss.view(batch_size, -1) * square_weights_by_sample.unsqueeze(1)
loss = loss.sum() / x_counts.sum()
```
